### ISSUE TITLE
Tests broken on timezone Netherlands

### ISSUE THREAD

**Issue by sailsbot** [balderdashy/sails#4012]
_From @dmarcelino on April 20, 2015 13:19_

As uncovered by @RWOverdijk, the adapter tests are broken when ran from the Netherlands timezone. We are not 100% sure if the issue lies in sails-postgresql or waterline-adapter-tests but it seems to be in the former since only sails-postgresql seems to be affected by it.

To expose this, I've created the timezone_bug  branch (https://github.com/balderdashy/sails-postgresql/commit/a5e42e2f4c973d6eb652d82b0a6d48c48b991e87) which changes travis timezone to NL and you can see the build breaks on 10 `date` related tests:
https://travis-ci.org/balderdashy/sails-postgresql/jobs/59237954#L263

I've also done a similar change to the waterline-adapter-tests (https://github.com/balderdashy/waterline-adapter-tests/commit/696178fc9db1e4857217f889a5b6bf6be836b34f) **integration** tests that run across all official adapters and **only sails-postgresql** broke these `date` tests: https://travis-ci.org/balderdashy/waterline-adapter-tests/jobs/59227210#L318

_Copied from original issue: balderdashy/sails-postgresql#153_

**Comment by sailsbot** [comment]
_From @dmarcelino on April 20, 2015 13:42_

As reported by @Zax in balderdashy/waterline#958, this can also be observed during migrations:

> Strange bug, 
> i configure a model with 'date' field and i observed this field on database (postgres). After all restart (sails lift) the field is decreased by one day, on develelop env (migration set on 'alter').
> If i switch env on production (migration set on 'save'), all right.

**Comment by sailsbot** [comment]
_From @RWOverdijk on April 20, 2015 17:48_

I found that passing in a UTCString to .find() "fixes" the problem.

I'm now looking into why this is a thing. It seems like there are two different methods of working with dates, and somewhere it's not doing something with the criteria (I was able to solve this but did so in a disgusting place. I'm looking for a nicer spot to put the fix.

**Comment by sailsbot** [comment]
_From @RWOverdijk on April 21, 2015 11:26_

Currently, we insert as UTC for postgres and fetch as local timezone for all. MySQL also inserts for local timezone. This causes bugs. Especially when trying to fix one of both. I can fix it, but I need to know what our stance on the matter will be.

@particlebanana I think we should pick a strategy and implement it. We can either:
- Store exactly what the user gives us, and return exactly what we stored.
- Store what the user gives us in UTC, and return in local timezone (Date.UTC()).

We should always pick strategy one for dates. However, DateTime and Time require, in my opinion, the second approach. That way it's safe to migrate data between timezones.

**Comment by sailsbot** [comment]
_From @dmarcelino on April 21, 2015 11:33_

cc: @tjwebb, @devinivy

**Comment by sailsbot** [comment]
_From @devinivy on April 21, 2015 12:31_

I like the second option the most.  Keeping that in mind, certain databases have very specific ideas about each of their time types.  For example,

> MySQL converts `TIMESTAMP` values from the current time zone to UTC for storage, and back from UTC to the current time zone for retrieval. (This does not occur for other types such as `DATETIME`.)

So first-off, let's make sure the types used by the adapters are in line with the types used by the databases.  In general, though, I think storing dates in UTC then retrieving to local time is our best bet.  It may be worth having a conversation around waterline having its own configurable timezone options to keep the apps portable.  Lastly, let's make sure data remains consistent for current users!  If my app currently stores dates in EST, I wouldn't want that to change with a release.

**Comment by sailsbot** [comment]
_From @RWOverdijk on April 21, 2015 13:2_

@devinivy If we made such a choice, it'd probably be either a major version up or a configurable value that defaults to a BC value.

But we do have to pick one. Seeing how adapters behave differently, we should either pick a default or embrace their difference. But I thought that waterline was all about unifying the adapters.

**Comment by sailsbot** [comment]
_From @devinivy on April 21, 2015 13:36_

I'd expect a minor version up, given that fixing this change doesn't imply that the adapter has hit stability implied by 1.0.  Breaking changes typically go into the minor version when the major verson is zero.  Also, did you mean that the configurable value could default to a _non-_ breaking change value?

I'm only saying that the adapter needs to use a type that caters to the job it's trying to do.  It's important to choose the column type in a way that it's actually possible for waterline to unify the adapters.  For example, it's important to take into account this sort of conversion that MySQL's `TIMESTAMP` does on its own while writing an adapter to conform to the unification of types that waterline's goal is to provide.  I think that each waterline date type needs to have a more specific role and behavior than is currently noted in the docs.

**Comment by sailsbot** [comment]
_From @dmarcelino on April 21, 2015 13:39_

@devinivy, regarding the different date times role in waterline, there was some discussion about it on https://github.com/balderdashy/anchor/pull/69#issuecomment-94057805. It's definitely something that should be clear and transparent.

**Comment by sailsbot** [comment]
_From @dmarcelino on April 24, 2015 16:54_

BTW, we should also take into account balderdashy/waterline#528 raised by @leedm777:

> When Cast.prototype.date parses a string, simple dates are parse at midnight UTC. But when the string is converted back into a date for putting in the database, it's formatted as local time. For anyone west of UTC, the date portion of the string is one day too early.

**Comment by sailsbot** [comment]
_From @bmwertman on May 18, 2015 1:2_

@dmarcelino posted a reply to my [Stack OverFlow post](http://stackoverflow.com/questions/30278990/why-does-my-post-request-to-postgresql-db-save-my-utc-datetime-as-a-local-time) regarding this issue. I'm referencing it here in case anyone else in on this discussion can shed some additional light on the problem. @RWOverdijk approximately how far out do you think we are talking about before a fix might be put in place?

**Comment by sailsbot** [comment]
_From @fundon on May 20, 2015 9:31_

**Please check this example**:

I also disabled createdAt and updatedAt.

``` js
  autoCreateAt: false,
  autoUpdateAt: false,

  attributes: {
    created_at: {
      type: 'datetime',
      notNull: true,
      defaultsTo: function() {
        return (new Date()).toISOString()
      }
    },

    updated_at: {
      type: 'datetime',
      notNull: true,
      defaultsTo: function() {
        return (new Date()).toISOString()
      }
    },
  }
```

Will get 2015-05-10 09:33:11.037000 when using timestamp without time zone type, so new Date(' 2015-05-10 09:33:11.037000') is wrong. That will be parsed by local time zone.

The right is `new Date('2015-05-10 09:33:11.037000+00:00')` or`new Date('2015-05-10 09:33:11.037000Z')`.

**Comment by sailsbot** [comment]
Thanks for posting, @dmarcelino.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
_From @dmarcelino on September 25, 2015 9:14_

@sailsbot, this one is a big bug with real consequences for people with timezone sensitive applications. As stated in my first post it has been reproduced in travis by changing timezone to NL: https://travis-ci.org/balderdashy/sails-postgresql/jobs/59237954#L263

**Comment by sailsbot** [comment]
Thanks for posting, @dmarcelino.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
_From @RWOverdijk on October 26, 2015 14:3_

Ping. Otherwise the bot will re-re-re-reclose

**Comment by sailsbot** [comment]
_From @fundon on October 26, 2015 18:29_

@RWOverdijk https://github.com/fundon/fundon.github.io/issues/21, Maybe it helps you.

**Comment by sailsbot** [comment]
_From @particlebanana on January 25, 2016 23:34_

@RWOverdijk based on the discussion what do we need to do here? If by local time you mean whatever is set on the server where the code is run then that can cause just as many issues.

I've always just saved integer timestamps and translated that to a datetime on the front-end. This is tough because of the auto-generated timestamps. I haven't looked into what other ORM's do.

**Comment by sailsbot** [comment]
_From @RWOverdijk on January 26, 2016 2:35_

@particlebanana A breaking change, but I'd say store as utc.

**Comment by sailsbot** [comment]
_From @RWOverdijk on March 26, 2016 22:14_

I'd like to re-bump this. Running into the same issue.

**Comment by sailsbot** [comment]
_From @cjnqt on September 13, 2016 18:58_

Any news about this?!

**Comment by sailsbot** [comment]
_From @gsdatta on October 5, 2016 23:50_

@cjnqt Doesn't look there's been activity on the timezone issue, it was a dealbreaker for me and even though I found a fix (https://github.com/sailsjs/anchor/pull/69) for the ORM, consensus wasn't reached on the best way to deal with it. I feel that timezone bugs should be prioritized as it is a critical component for an ORM which can be used in prod.

**Comment by sailsbot** [comment]
_From @sgress454 on March 1, 2017 22:40_

Looks like tests for sails-postgresql 1.0 pass w/ the timezone set to Amsterdam, probably because timestamps now default to integers instead of ISO Date strings.

**Comment by sgress454** [comment]
🎉Closing as this works in the official 1.0 release.
